### PR TITLE
feat: Huff tests

### PR DESCRIPTION
**Pull request by clabby**
## Overview

Adds a stripped down test suite to huff-rs, mainly for gas profiling and creating assertions.
Peek 2022-08-07 00-56

### Usage
CLI:
`huffc <contract_path> -t [-f <list|table|json>] [-m <TEST_NAME>]`

Test syntax:
```
#[calldata("0x0000000000000000000000000000000000000000000000000000000000000020"), value(0x01)]
#define test MY_TEST() = takes (0) returns (0) {
    // test code
}

#define test MY_TEST_2() = takes (0) returns (0) {
    // test code
}

#define macro MY_MACRO() = takes (0) returns (0) {
    MY_TEST() // Tests cannot be invoked- this will throw an error.
}
```

### Checklist
- [x] Create a test runner that utilizes revm
  - [x] Tests for the runner
  - [x] Configurable test environment (gas limit, chain id, hardfork, call value, etc.)
  - [ ] Simple input fuzzer (? - not necessarily needed, might be nice for fuzzed gas profiling.)
- [x] Implement test macros into the lexer / parser
  - [x] Restrict regular macros / functions from invoking test macros
- [x] Add a subcommand / flag to the CLI to run the test suite
- [x] Clean up
  - [x] Proper errors
  - [x] README

### CODE REVIEW

**Comment by ncitron** [comment]
This is fantastic! One thought I have is whether it is worth it to work on our own test runner, or to just hijack foundry to handle it. A custom test runner is convenient, but is hard work to implement and maintain. Another solution is to add a command line flag that compiles each of the test functions as if they were `MAIN` to generate a list of test contracts. We could then call these functions from foundry to run the tests (with failing tests reverting). This even adds compatibility with fuzz tests, since you can write your test functions to do a quick `CALLDATALOAD` and write a foundry fuzz test that calls the test contract with the randomly generated values.

What do you think?

**Comment by clabby** [comment]
> This is fantastic! One thought I have is whether it is worth it to work on our own test runner, or to just hijack foundry to handle it. A custom test runner is convenient, but is hard work to implement and maintain. Another solution is to add a command line flag that compiles each of the test functions as if they were `MAIN` to generate a list of test contracts. We could then call these functions from foundry to run the tests (with failing tests reverting). This even adds compatibility with fuzz tests, since you can write your test functions to do a quick `CALLDATALOAD` and write a foundry fuzz test that calls the test contract with the randomly generated values. A couple helpers in foundry-huff could make this process even easier.
> 
> What do you think?

Thanks for the feedback! Definitely all great points. The idea behind this at the moment is primarily to give the ability to make quick assertions and to provide a more accurate gas profiler for the devs. Not necessarily to replace `foundry-huff`, but to be complimentary to it. It could definitely grow into a more fleshed out test suite, but like you said, I think there's no point in reinventing the wheel in that regard. Foundry is already amazing and fits almost all needs :smile: 

1. Using a custom test runner was not my first choice, but after some looking into using foundry's `evm` module, it was heavy and implemented a lot of features we either don't need or that we'd have to add in other parts of the codebase to use. With that, it seemed easier to create a stripped down version that only executes arbitrary bytecode (compiled tests, as if they were the `MAIN` macro) and tells us the amount of gas it consumed, whether or not it reverted, and the return/revert data. What's currently there should be over half of what we need for that, and my thinking is that it shouldn't be too much to maintain considering the lack of features.
2. The idea for a cli flag to create test contracts is definitely interesting- I like the idea. I don't know how much that would help, though, since that is effectively what test wrappers (as seen in Huffmate). Also, there's still the problem of foundry reporting incorrect gas costs. In terms of fuzz testing, I think that for most applications, it'd be better to use a wrapper contract with foundry-huff. We can definitely add fuzzing inputs to Huff tests, but I figure people would most commonly use them for average gas cost based on a wide range of inputs.

Definitely down to open the discussion- the direction is fairly loose at the moment. Maybe we should ask the discord & come to some consensus on what this should be?

**Review by abigger87 — CHANGES_REQUESTED** [review]
Just a few nits

File: huff_core/src/lib.rs
Comment (clabby) [reply huff_core/src/lib.rs]: Removed :+1:

File: huff_core/src/lib.rs
Comment (clabby) [reply huff_core/src/lib.rs:269]: Amazing :crown: 

Should we wait until #202 is merged so that I can parallelize this?

Comment (abigger87) [reply huff_core/src/lib.rs:269]: should be done in 202 :)

and 202 is meant to be merged into this branch :)

Comment (clabby) [reply huff_core/src/lib.rs:269]: > should be done in 202 :)
> 
> and 202 is meant to be merged into this branch :)

Ohh whoops, didn't see that part. Will review it now :pray:

**Pull request merged**
